### PR TITLE
Change Shewchuk Orientation filter to Ozaki et al. filter.

### PR DESCRIPTION
**Pull request by tinko92**
This commit changes the current Shewchuk-based (notably, the original code doesn't use Shewchuk's constant which is significantly narrower) to the filter published by Ozaki et al. in https://doi.org/10.1007/s10543-015-0574-9 . It is tighter, so it has strictly fewer misses and it has fewer and better predictable branches and fewer instructions so it is theoretically and practically faster (the referenced paper has benchmarks).

It does not come with new tests because it should not change any observable behavior by design except possibly speed, which is not tested, so all tests testing the current predicate are tests for the patched implementation.

### CODE REVIEW

**Comment by dr-jts** [comment]
Where does the value 3.3306690621773714e-16 come from?  I don't see it in the Ozaki paper (source from [here](https://www.tuhh.de/ti3/paper/rump/OzBueOgOiRu15.pdf)).

**Comment by dr-jts** [comment]
See also https://github.com/libgeos/geos/pull/1184

**Comment by dr-jts** [comment]
I wonder if this solves #750?

**Comment by tinko92** [comment]
> I wonder if this solves #750?
> 
> Specifically, what is the orientation index computed for
> 
> ```
> LINESTRING (0 100, 3 106.3246)
> 
> POINT (1 102.1082)
> ```
> 
> Currently this is computed as -1, but by inspection it should be 0 (collinear).
> 
> If so, this case can be added as a unit test.

-1 is correct for this test case if the numbers are read as "nearest double to 106.3246" and "nearest double to 102.1082", which is how a compiler or text parser will read them, here is a visualization of the double values close to that: 
image

Blue represents -1, Red represents 1, Green represents 0. Notably, there is no green in this picture because the exact colinear point at x=1 is not representable in double precision. The pixels are not quadratic because at 1 and 100, the exponents of the x and y coordinates are different.

I can see how this is an unexpected result for library users, though. I see three possible solutions to solve this:

- Use this exact predicate for construction to guarantee algorithmic correctness.
- Use another rounding predicate just for simple point to linear tests to be more in line with user expectation where inconsistent individual predicate results cannot mess up the consistency of a more complicated datastructure like a computed intersection polynomial or a triangulation. In this case in the filter failure case 0 should be returned and the constant should be changed from ~3e-16 to roughly ~5e-16 (can try to compute the exact constant later) to guarantee that including the initial rounding error from a decimal input, "collinear up to rounding" points are recognised as such after the accumulated error of the determinant computation. This would also find some points collinear whose decimal representations are not collinear and will produce subtly inconsistent results between different algorithm.
- Use a decimal number type and extended precision arithmetic at the cost of a severe performance penalty (I'd expect at least 1 order of magnitude slower operations for everything where predicate computation cost is not negligible like triangulations).

**Comment by dr-jts** [comment]
@tinko92 Great analysis, thanks.  It sounds like there is no easy/fast solution to this problem.  That's not a surprise, given the well-known issues of geometric robustness when using floating-point.  We'll just have document this situation, and ideally provide tolerance-based algorithms as a work-around.

**Comment by grootstebozewolf** [comment]
I've machine-checked soundness proofs for both the Shewchuk-based 
and Ozaki γ₂ filter thresholds in Rocq/Flocq:

  b64_orient_sign_stage_d_sound (Stage D, expansion arithmetic)
  b64_ozaki_filter_sound (Ozaki γ₂ threshold, this PR's approach)

Both theorems confirm the same interpretation documented in this 
discussion: "correct sign of the exact determinant of the actual 
double inputs", not "collinear up to decimal rounding".

Corpus: https://github.com/grootstebozewolf/NetTopologySuite.Proofs
Relevant file: theories-flocq/Orient_b64_Ozaki.v

**Comment by micycle1** [comment]
A relevant 2023 paper, if this topic is revisited: [s10543-023-00975-x.pdf](https://github.com/user-attachments/files/29918531/s10543-023-00975-x.pdf).

> More recently, Ozaki et al. developed an improved static filter as well as a new semi-static filter for the 2D orientation predicate, where the latter also handles floating-point exceptions such as overflow and underflow. Their filters, however, are not designed for arbitrary polynomial predicates.

image

**Comment by grootstebozewolf** [comment]
Thanks for this — differential check from the NetTopologySuite.Proofs side (precision / orientation lane).

We mirrored **three** Stage A shapes in pure double and gated them against an independent extracted exact orient (`ORIENT_EXACT` / `b64_orient2d_exact`). We did **not** re-run the full Maven suite or claim bit-identity of the DD fallback with the oracle.

### Method
- **Oracle GT:** `ORIENT_EXACT` (full-plane exact sign).
- **JTS-master mirror:** current `orientationIndexFilter` — Shewchuk-style detsum + `DP_SAFE_EPSILON = 1e-15`.
- **Ozaki mirror (this PR):** `errbound = |detleft + detright| * 3.3306690621773724e-16`, CERTAIN iff `|det| >= errbound`.
- **Corpus Shewchuk Stage A:** detsum + `ccwerrboundA = (3+16ε)ε` (ε = 2⁻⁵³).

Write-up + table: [jts-1093-orient-lane](https://github.com/grootstebozewolf/NetTopologySuite.Proofs/blob/main/docs/jts-1093-orient-lane-2026-08.md) (`oracle/jts1093_orient_vectors.txt`, 45 vectors).

### Results
| Check | Result |
|---|---|
| Certified pin failures (EXPECTED set) | **0** |
| Ozaki **CERTAIN** vs exact sign conflicts | **0** |
| Corpus Shewchuk **CERTAIN** vs exact conflicts | **0** |
| JTS-master **CERTAIN** vs exact conflicts | **0** |
| Ozaki UNCERTAIN / Shewchuk UNCERTAIN / master UNCERTAIN | **10 / 10 / 14** |
| Ozaki CERTAIN while master UNCERTAIN | **4** |
| Master CERTAIN while Ozaki UNCERTAIN | **0** |
| Ozaki ↔ corpus Shewchuk certainty splits | **0** |

### “Fewer misses” band (same-sign products)

```text
a = (0, 0),  b = (1, 0.5),  q = (2, 1 + h)
```

| h | Exact | Master (1e-15) | Ozaki | Shewchuk ccwerrboundA |
|---|---|---|---|---|
| 1e-14 | ± | CERTAIN | CERTAIN | CERTAIN |
| **1e-15, 2e-15** | ± | **UNCERTAIN** | **CERTAIN** | **CERTAIN** |
| 5e-16, ~3.3e-16 | ± | UNCERTAIN | UNCERTAIN | UNCERTAIN |

So on this table the PR’s claim holds in miniature: the master constant escalates earlier; Ozaki commits correctly in the gap and never committed a wrong sign. Ozaki’s published double sits next to true Shewchuk `ccwerrboundA` (they never split certainty here). Deep collinear / zero-det cases still correctly force UNCERTAIN → DD, as designed.

### Bottom line
From an independent exact-orient gate, this filter swap looks **sound and slightly tighter** than master, and **aligned with** a Shewchuk-constant Stage A rather than the historical `1e-15` guard. Happy to re-run if you add more near-collinear fixtures to the suite.

*(Affiliation: [NetTopologySuite.Proofs](https://github.com/grootstebozewolf/NetTopologySuite.Proofs) — formal precision/predicate companion, not a JTS committer review.)*
